packages/flutter/test/services/text_input_test.dart

```diff
@@ -90,3 +143,79 @@ void main() {
     });
   });
 }
+
+class FakeTextInputClient implements TextInputClient {
+  const FakeTextInputClient();
+
+  TextInputConfiguration get configuration => const TextInputConfiguration();
+
+  @override
+  void performAction(TextInputAction action) => throw UnimplementedError();
+
+  @override
+  void updateEditingValue(TextEditingValue value) => throw UnimplementedError();
+  @override
+  void updateFloatingCursor(RawFloatingCursorPoint point) => throw UnimplementedError();
+}
+
+class FakeTextChannel implements MethodChannel {
+  FakeTextChannel(this.outgoing) : assert(outgoing != null);
+
+  Future<void> Function(MethodCall) outgoing;
+  Future<void> Function(MethodCall) incoming;
+
+  List<MethodCall> outgoingCalls = <MethodCall>[];
+
+  @override
+  BinaryMessenger get binaryMessenger => throw UnimplementedError();
+
+  @override
+  MethodCodec get codec => const JSONMethodCodec();
+
+  @override
+  Future<List<T>> invokeListMethod<T>(String method, [dynamic arguments]) => throw UnimplementedError();
+
+  @override
+  Future<Map<K, V>> invokeMapMethod<K, V>(String method, [dynamic arguments]) => throw UnimplementedError();
+
+  @override
+  Future<T> invokeMethod<T>(String method, [dynamic arguments]) {
+    final MethodCall call = MethodCall(method, arguments);
+    outgoingCalls.add(call);
+    return outgoing(call);
+  }
+
+  @override
+  String get name => 'flutter/textinput';
+
+
+  @override
+  void setMethodCallHandler(Future<void> Function(MethodCall call) handler) {
+    incoming = handler;
+  }
+
+  @override
+  void setMockMethodCallHandler(Future<void> Function(MethodCall call) handler)  => throw UnimplementedError();
+
+  void validateOutgoingMethodCalls(List<MethodCall> calls) {
+    expect(outgoingCalls.length, calls.length);
+    bool hasError = false;
+    for (int i = 0; i < calls.length; i++) {
+      final ByteData outgoingData = codec.encodeMethodCall(outgoingCalls[i]);
+      final ByteData expectedData = codec.encodeMethodCall(calls[i]);
+      final String outgoingString = utf8.decode(outgoingData.buffer.asUint8List());
+      final String expectedString = utf8.decode(expectedData.buffer.asUint8List());
+
+      if (outgoingString != expectedString) {
+        print('''Index $i did not match:
```

Comment:
This would format a little nicer as:

```dart
print(
    'Index $i did not match:\n'
    '  actual: ${outgoingCalls[i]}\n'
    '  expected: ${calls[i]}'
);
```

Reply:
Agreed, done.